### PR TITLE
Migrate PipelineResource in their own package… 📦

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

Extracts PipelineResource in their own package while still being exposed in `tekton.dev`. **This is require to support embedded PipelineResource in `v1alpha2` with PipelineResource still being alpha (aka `v1alpha1`)**.

- This adds a `pkg/apis/resource/v1alpha1` package which holds `PipelineResource` struct and related.
- This adds a `pkg/client/resource` package (in addition to the `pkg/client` package) for the `PipelineResource` generated clients, lister, … code.

<del>… and API group (`resource.tekton.dev`). This is mainly done to
separate the `PipelineResource` from the rest of the API *and* remove
the dependency cycle that we get by keeping `PipelineResource` in
alpha.</del>

<del>This doesn't change yet the current webhook and controller, and does
not expose `PipelineResource.resource.tekton.dev` (current exposed are
`PipelineResource.tekton.dev`).</del>

<del>We can expose this and support this once we expose `v1alpha2` (or
`v1beta1`). We will need to auto-convert `PipelineResource.tekton.dev`
to `PipelineResource.resource.tekton.dev`, the easiest way would be to
have a controller handling that.</del>

This comes from this comment https://github.com/tektoncd/pipeline/pull/1725#discussion_r356158794 :angel: .

This is an experiment, seeing if it make sense. Without this, it is very difficult to support embedded PipelineResource in `v1alpha2` (or I am missing something really obvious :stuck_out_tongue_closed_eyes: )
/hold

/cc @bobcatfish 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: cmd/webhook/main.go [hunk 1]
```diff
@@ -98,12 +98,13 @@ func main() {
...
 	resourceHandlers := map[schema.GroupVersionKind]webhook.GenericCRD{
 		v1alpha1.SchemeGroupVersion.WithKind("Pipeline"):         &v1alpha1.Pipeline{},
-		v1alpha1.SchemeGroupVersion.WithKind("PipelineResource"): &v1alpha1.PipelineResource{},
 		v1alpha1.SchemeGroupVersion.WithKind("Task"):             &v1alpha1.Task{},
 		v1alpha1.SchemeGroupVersion.WithKind("ClusterTask"):      &v1alpha1.ClusterTask{},
 		v1alpha1.SchemeGroupVersion.WithKind("TaskRun"):          &v1alpha1.TaskRun{},
 		v1alpha1.SchemeGroupVersion.WithKind("PipelineRun"):      &v1alpha1.PipelineRun{},
 		v1alpha1.SchemeGroupVersion.WithKind("Condition"):        &v1alpha1.Condition{},
+		v1alpha1.SchemeGroupVersion.WithKind("PipelineResource"): &v1alpha1.PipelineResource{},
+		// resourcev1alpha1.SchemeGroupVersion.WithKind("PipelineResource"): &resourcev1alpha1.PipelineResource{},
```
Comment (bobcatfish) [inline cmd/webhook/main.go]: Maybe accidentally included? Or if this is the line that exposes the API and we want to uncomment it later, maybe a comment above with an issue number explaining?

File: pkg/apis/resource/v1alpha1/pipelineresource_validation_test.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if v1alpha1.AllowedStorageType(tt.storageType) != tt.want {
+				t.Errorf("PipleineResource.allowedStorageType should return %t, got %t for %s", tt.want, !tt.want, tt.storageType)
+			}
+		})
+	}
+
+}
```
Comment (bobcatfish) [inline pkg/apis/resource/v1alpha1/pipelineresource_validation_test.go]: It seems like some of these files are duplicated vs. moved? e.g. would https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipelineresource_validation_test.go still exist after this PR?

**Comment by vdemeester** [comment]
> I think it makes a lot of sense to move the code, but I'm a bit confused by why we want to add a new API group - looking at our other types, and even looking at triggers types such as [trigger bindings](https://github.com/tektoncd/triggers/blob/27b3960dc647e392ff110090150406beea3a04b2/config/300-triggerbinding.yaml#L20-L21) they are still in the `tekton.dev` group (or maybe this is a mistake and the group should be different?), so I don't quite understand why we want pipelineresources to be an exceptional case thinking

Right, on `triggers`, I would really like having them in a different group (make it easier to *detect* the precense of trigger without having to know which types it exposes — right now.. ` | grep tekton.dev` returns pipeline and triggers :stuck_out_tongue:).

The main problem I've got, so far, is that I did not manage to make the codegenerator(s) work with two distinct package serving the exact same api group (the clients are from one package only making the thing useless). I am still trying to figure out how to do that, as it would definitely simplify that pull request.

That said, I think we may discuss `apiGroup` in pipeline and in tekton in general :angel:.

**Comment by bobcatfish** [comment]
> Right, on triggers, I would really like having them in a different group (make it easier to detect the precense of trigger without having to know which types it exposes — right now.. | grep tekton.dev returns pipeline and triggers stuck_out_tongue).

Ah okay, so in that case maybe we should have `pipeline.tekton.dev` and `triggers.tekton.dev` - i.e. mirror our repo layout?

I'm still not sure it would make sense to put resources in it's own group :thinking: on the other hand that might fit well with some of the plans @sbwsg and I have been scheming :smiling_imp: 

> The main problem I've got, so far, is that I did not manage to make the codegenerator(s) work with two distinct package serving the exact same api group (the clients are from one package only making the thing useless). I am still trying to figure out how to do that, as it would definitely simplify that pull request.

I think it'd be better if we could get this to work, esp. if the main problem is that the tool doesn't work. If you want to write up an issue and get any of the rest of us to take a stab at it just say the word!! happy to help :D

**Comment by vdemeester** [comment]
@bobcatfish should be fixed now, no more different apiGroup.

**Comment by vdemeester** [comment]
@bobcatfish @sbwsg this should be ready for review :angel:

File: hack/update-codegen.sh [hunk 3]
```diff
@@ -37,10 +42,15 @@ ${GOPATH}/bin/deepcopy-gen \
...
 # Knative Injection
+bash ${REPO_ROOT_DIR}/hack/generate-knative.sh "injection" \
+  github.com/tektoncd/pipeline/pkg/client/resource github.com/tektoncd/pipeline/pkg/apis \
+  "resource:v1alpha1" \
+  --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt
 bash ${REPO_ROOT_DIR}/hack/generate-knative.sh "injection" \
   github.com/tektoncd/pipeline/pkg/client github.com/tektoncd/pipeline/pkg/apis \
   "pipeline:v1alpha1,v1alpha2" \
   --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt
+GOFLAGS="${OLDGOFLAGS}"
```
Comment (bobcatfish) [inline hack/update-codegen.sh]: is it possible to add some comments around the enw lines in this script explaining what they do? and/or explain in the commit message? iirc from the previous review, it was not obvious what you had to do to make this work without a different apigroup, might be useful to record how this ended up working

Comment (vdemeester) [reply hack/update-codegen.sh]: ah this comes from https://github.com/tektoncd/pipeline/pull/1783.. forgot to mention that :joy: :sweat:

Comment (vdemeester) [reply hack/update-codegen.sh]: I need to update comments on the `bash …` calls though, to explain why there is 2 indeed

Comment (bobcatfish) [reply hack/update-codegen.sh]: kk! happy for that to be in follow up PR :)

**Comment by bobcatfish** [comment]
/lgtm

thanks again for all your hard work @vdemeester 🙏 😭 🙇‍♀

**Pull request merged**